File: Java/e_learn/src/main/java/gr/aueb/elearn/chapter14/AbstractAnimal.java
package gr.aueb.elearn.chapter14;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
public abstract class AbstractAnimal {
    private Long id;
    private String name;
    private Double age;
 
    // public API
    public abstract void speak();

    public void eat() {
        System.out.println("Animal is eating!");
    }

    @Override
    public abstract String toString();
}

File: Csharp/SebAuebAug2022/AbstractClassesApp/AbstractAnimal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AbstractClassesApp
{
    internal abstract class AbstractAnimal
    {
        public int Id { get; set; }
        public string? Name { get; set; }
        public double Age { get; set; }

        public abstract void Speak();

        public virtual void Eat()
        {
            Console.WriteLine("The animal choked on the food that it was eating :-)");
        }

        public abstract override string ToString();
    }
}
